### PR TITLE
Gimbal Lock in Quaternion -> Euler angles conversion

### PR DESCRIPTION
**Pull request by bathal1**
This PR introduces a fix to the quaternion to Euler angles conversion code to handle gimbal lock properly. A unit test (on the python side) has also been added, with conversion values determined in Matlab.

The reverse conversion Euler -> Quaternion has also been implemented.

### CODE REVIEW

File: enoki/matrix.py [hunk 1]
```diff
@@ -267,30 +267,46 @@ def matrix_to_quat(m):
 
 
 def quat_to_euler(q):
-    q_y_2 = _ek.sqr(q.y)
+    name = _ek.detail.array_name('Array', q.Type, [3], q.IsScalar)
+    module = _modules.get(q.__module__)
+    Array3f = getattr(module, name)
 
+    sinp = 2 * _ek.fmsub(q.w, q.y, q.z * q.x)
+    gimbal_lock = _ek.abs(sinp) > 1.0 - 1e-5
```
Comment (Speierers) [inline enoki/matrix.py:275]:
```suggestion
    gimbal_lock = _ek.abs(sinp) > (1.0 - 1e-5)
```
Just to clarify your intentions 😄

File: enoki/matrix.py [hunk 2]
```diff
@@ -267,30 +267,46 @@ def matrix_to_quat(m):
...
+
+    # roll (x-axis rotation)
+    q_y_2 = _ek.sqr(q.y)
     sinr_cosp = 2 * _ek.fmadd(q.w, q.x, q.y * q.z)
     cosr_cosp = _ek.fnmadd(2, _ek.fmadd(q.x, q.x, q_y_2), 1)
-    roll = _ek.atan2(sinr_cosp, cosr_cosp)
+    roll = _ek.select(gimbal_lock, 2*_ek.atan2(q.x, q.w), _ek.atan2(sinr_cosp, cosr_cosp))
```
Comment (Speierers) [inline enoki/matrix.py:281]:
```suggestion
    roll = _ek.select(gimbal_lock, 2 * _ek.atan2(q.x, q.w), _ek.atan2(sinr_cosp, cosr_cosp))
```

File: enoki/matrix.py [hunk 3]
```diff
@@ -267,30 +267,46 @@ def matrix_to_quat(m):
...
 
     # pitch (y-axis rotation)
-    sinp = 2 * _ek.fmsub(q.w, q.y, q.z * q.x)
-    if (_ek.abs(sinp) >= 1.0):
-        pitch = _ek.copysign(0.5 * _ek.Pi, sinp)
-    else:
-        pitch = _ek.asin(sinp)
+    pitch = _ek.select(gimbal_lock, _ek.copysign(0.5*_ek.Pi, sinp), _ek.asin(sinp))
```
Comment (Speierers) [inline enoki/matrix.py:284]:
```suggestion
    pitch = _ek.select(gimbal_lock, _ek.copysign(0.5 * _ek.Pi, sinp), _ek.asin(sinp))
```

File: enoki/matrix.py (hunk 1 above)
Comment (Speierers) [reply enoki/matrix.py:275]: Where does `1e-5` come from? Could you use `ek.Epsilon(type(sinp))` instead?

File: enoki/matrix.py [hunk 4]
```diff
@@ -267,30 +267,46 @@ def matrix_to_quat(m):
...
+    name = _ek.detail.array_name('Quaternion', angles.Type, [4], angles.IsScalar)
+    module = _modules.get(angles.__module__)
+    Quat4f = getattr(module, name)
+
+    cr = _ek.cos(angles.x / 2.0)
+    sr = _ek.sin(angles.x / 2.0)
+    cp = _ek.cos(angles.y / 2.0)
+    sp = _ek.sin(angles.y / 2.0)
+    cy = _ek.cos(angles.z / 2.0)
+    sy = _ek.sin(angles.z / 2.0)
```
Comment (Speierers) [inline enoki/matrix.py:303]:
```suggestion
    angles /= 2.0
    sr, cr = _ek.sincos(angles.x)
    sp, cp = _ek.sincos(angles.y)
    sy, cy = _ek.sincos(angles.z)
```

File: enoki/matrix.py (hunk 1 above)
Comment (Speierers) [reply enoki/matrix.py:275]: Or even `_ek.OneMinusEpsilon(type(sinp))`

File: include/enoki/quaternion.h [hunk 5]
```diff
@@ -233,28 +233,43 @@ template <typename Value>
 Array<Value, 3> quat_to_euler(const Quaternion<Value> &q) {
     // https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
 
-    Value q_y_2 = sqr(q.y());
+    Value sinp = 2 * fmsub(q.w(), q.y(), q.z() * q.x());
+    Mask gimbal_lock = abs(sinp) > 1.f - 1e-5;
```
Comment (Speierers) [inline include/enoki/quaternion.h:237]: Please use `ek::OneMinusEpsilon<Value>`

File: include/enoki/quaternion.h [hunk 6]
```diff
@@ -233,28 +233,43 @@ template <typename Value>
...
+template <typename Value>
+Quaternion<Value> euler_to_quat(const Array<Value, 3> &a) {
+    // https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+
+    Value cr = cos(a.x / 2.0f);
+    Value sr = sin(a.x / 2.0f);
+    Value cp = cos(a.y / 2.0f);
+    Value sp = sin(a.y / 2.0f);
+    Value cy = cos(a.z / 2.0f);
+    Value sy = sin(a.z / 2.0f);
```
Comment (Speierers) [inline include/enoki/quaternion.h:265]: Same as in python code (using `ek::sincos()`)

File: tests/python/test_matrix.py [hunk 7]
```diff
@@ -260,3 +259,23 @@ def to_dest(a):
...
     ])
     assert(m == Matrix44f(to_dest(m)))
+
+@pytest.mark.parametrize("package", [ek.scalar])
+def test15_quat_to_euler(package):
+    Quaternion4f, Array3f, Float = package.Quaternion4f, package.Array3f, package.Float
+    prepare(package)
+
+    # Gimbal lock at +pi/2
+    q = Quaternion4f(0, 1/np.sqrt(2), 0, 1/np.sqrt(2))
```
Comment (Speierers) [inline tests/python/test_matrix.py:269]:
```suggestion
    q = Quaternion4f(0.0, 1.0 / np.sqrt(2), 0.0, 1.0 / np.sqrt(2))
```

And please do the same for the rest of this function

File: tests/python/test_matrix.py [hunk 8]
```diff
@@ -260,3 +259,23 @@ def to_dest(a):
...
+    assert(ek.allclose(ek.quat_to_euler(q), Array3f(0, np.pi/2, 0)))
+    # Gimbal lock at -pi/2
+    q = Quaternion4f(0, -1/np.sqrt(2), 0, 1/np.sqrt(2))
+    assert(ek.allclose(ek.quat_to_euler(q), Array3f(0, -np.pi/2, 0)))
+    # Quaternion without gimbal lock
+    q = Quaternion4f(0.15849363803863525, 0.5915063619613647, 0.15849363803863525,0.7745190262794495)
```
Comment (Speierers) [inline tests/python/test_matrix.py:275]:
```suggestion
    q = Quaternion4f(0.15849363803863525, 0.5915063619613647, 0.15849363803863525, 0.7745190262794495)
```

**Review by Speierers — COMMENTED** [review]
Thanks @bathal1 for this PR.

Here are a few minor comments. It looks great otherwise!

**Comment by Speierers** [comment]
Well done @bathal1 ! 🚀

**Comment by wjakob** [comment]
(Looks great 👍 )

File: include/enoki/quaternion.h [hunk 9]
```diff
@@ -233,28 +233,41 @@ template <typename Value>
...
+    Array<Value, 3> angles = a / 2.0f;
+    auto [sr, cr] = sincos(angles.x);
+    auto [sp, cp] = sincos(angles.y);
+    auto [sy, cy] = sincos(angles.z);
+
+    Value w = cr*cp*cy + sr*sp*sy;
+    Value x = sr*cp*cy - cr*sp*sy;
+    Value y = cr*sp*cy + sr*cp*sy;
+    Value z = cr*cp*sy - sr*sp*cy;
+    return Quaternion<Value>(x, y, z, w);
```
Comment (YunHsiao) [inline include/enoki/quaternion.h:269]: Just a quick note here, I think you've mixed different rotation orders for the conversions. The `euler_to_quat` uses `XYZ` order while `quat_to_euler` uses `ZYX` order. Just flip all the plus/minus signs in the above four lines would do.

Comment (Speierers) [reply include/enoki/quaternion.h:269]: @bathal1 could you take a look at this?

Comment (bathal1) [reply include/enoki/quaternion.h:269]: Hi @YunHsiao, do you have a reference for this? The computations for the `quat_to_euler` function have been derived from the 4 lines you mentioned, and we also have a functioning round-trip test comparing a Euler-angles rotation `e` with `quat_to_euler(euler_to_quat(e))` that breaks when applying your suggested change, so I find it hard to believe. The conversions we get from these are also consistent with matlab's `eul2quat` and `quat2eul` functions, that use the `ZYX` order by default.

Comment (YunHsiao) [reply include/enoki/quaternion.h:269]: Oh I'm so sorry I thought the components are declared in the order `xyzw`, but it's `wxyz`... The implementation is correct, my bad.

Comment (YunHsiao) [reply include/enoki/quaternion.h:269]: Prof. Shoemake has an excellent write up on this topic back in 1994, see Graphics Gem 4, Ch.III.5, p.222

Comment (bathal1) [reply include/enoki/quaternion.h:269]: Aha, great then! Thank you for the reference 🙂

**Pull request merged**